### PR TITLE
Accept Relationship business logic, api, tests; Notify Delegate business logic, api, tests; Creating identity automatically on discovery; Security Middleware Improvements; ForgeRock Middleware Improvements

### PR DESCRIPTION
**Pull request by dave-cheong**

### CODE REVIEW

File: backend/typescript/controllers/relationship.controller.ts [hunk 1]
```diff
@@ -49,8 +66,8 @@ export class RelationshipController {
         };
         validateReqSchema(req, schema)
             .then((req:Request) => this.relationshipModel.findPendingByInvitationCodeInDateRange(req.params.invitationCode, new Date()))
-            .then((model) => model.rejectPendingInvitation())
-            .then((model) => Promise.resolve({}))
+            .then((model) => model ? model.rejectPendingInvitation() : null)
```
Comment (sundriver) [inline backend/typescript/controllers/relationship.controller.ts]: why not just use an if statement, very hard to read the code under it.
if (mode) { ... return {} } else return null;

File: commons/RamAPI.ts [hunk 2]
```diff
@@ -252,7 +260,7 @@ export class AttributeDTO {
 export class RelationshipAddDTO {
     constructor(public relationshipTypeCode:string,
                 public subjectIdValue:string,
-                public delegate:IdentityDTO,
+                public delegate:CreateIdentityDTO,
```
Comment (sundriver) [inline commons/RamAPI.ts]: for DTOs we need interfaces all the way through, please look into Ram2 api

File: backend/typescript/controllers/relationship.controller.ts [hunk 3]
```diff
@@ -49,14 +51,53 @@ export class RelationshipController {
...
+        const schema = {
+            'invitationCode': {
+                notEmpty: true,
+                errorMessage: 'Invitation Code is not valid'
+            }
+        };
+        validateReqSchema(req, schema)
+            .then((req:Request) => this.relationshipModel.findPendingByInvitationCodeInDateRange(req.params.invitationCode, new Date()))
+            .then((model) => model ? model.rejectPendingInvitation() : null)
+            .then((model) => model ? Promise.resolve({}) : null)
```
Comment (sundriver) [inline backend/typescript/controllers/relationship.controller.ts]: could you also convert it to if or remove it

File: backend/typescript/tests/party.model.spec.ts [hunk 4]
```diff
@@ -1,80 +1,152 @@
...
 
-        try {
+        Seeder.verbose(false);
 
-            name1 = await NameModel.create({
-                givenName: 'John',
-                familyName: 'Smith'
-            });
+        Promise.resolve(null)
+            .then()
```
Comment (sundriver) [inline backend/typescript/tests/party.model.spec.ts]: empty then

File: backend/typescript/controllers/relationship.controller.ts (hunk 3 above)
Comment (dave-cheong) [inline backend/typescript/controllers/relationship.controller.ts]: Nah, we need this. It works and good. Forget this.

File: backend/typescript/tests/party.model.spec.ts (hunk 4 above)
Comment (dave-cheong) [inline backend/typescript/tests/party.model.spec.ts]: Removed

**Pull request merged**